aspnetcore/grpc/interceptors.md

```diff
@@ -0,0 +1,270 @@
+---
+title: gRPC interceptors on .NET
+author: erni27
+description: Learn how to use gRPC interceptors on .NET.
+monikerRange: '>= aspnetcore-3.0'
+ms.author:
+ms.date: 02/26/2022
+no-loc: [Home, Privacy, Kestrel, appsettings.json, "ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
+uid: grpc/interceptors
+---
+# gRPC interceptors on .NET
+
+By [Ernest Nguyen](https://github.com/erni27)
+
+Interceptors allow you to intercept incoming or outgoing gRPC requests. They offer a way to enrich the request processing pipeline. Since they are transparent for the user's application logic and once applied are triggered automatically, interceptors became a perfect solution for common cases like logging, monitoring, authentication, validation, etc.
+
+Interceptors can be implemented for both gRPC servers and clients by creating a class that inherits from `Interceptor` type.
+
+```csharp
+public class ExemplaryInterceptor : Interceptor
+{
+}
+```
+
+By default `Interceptor` base class doesn't do anything. Its virtual methods just passes the calls through. Thus, implementing your own interceptor comes down to overriding appropriate methods.
+
+## Client interceptors
+
+gRPC client interceptors intercept outgoing RPC invocations. They provide access to the sent request, the incoming response and the context for a client-side call.
+
+`Interceptor` methods to override for client:
+* `BlockingUnaryCall`- intercepts a blocking invocation of an unary RPC.
+* `AsyncUnaryCall` - intercepts an asynchronous invocation of an unary RPC.
+* `AsyncClientStreamingCall` - intercepts an asynchronous invocation of a client streaming RPC.
+* `AsyncServerStreamingCall` - intercepts an asynchronous invocation of a server streaming RPC.
+* `AsyncDuplexStreamingCall` - intercepts an asynchronous invocation of a bidirectional streaming RPC.
+
+> [!WARNING]
+> Although both `BlockingUnaryCall` and `AsyncUnaryCall` refer to unary RPCs, they can't be used interchangeably. A blocking invocation would not be intercepted by `AsyncUnaryCall` and an asynchronous one by `BlockingUnaryCall`.
+
+### Create a client gRPC interceptor
+
+The following code presents an example of intercepting an asynchronous invocation of a simple remote call.
+
+```csharp
+public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(
+    TRequest request,
+    ClientInterceptorContext<TRequest, TResponse> context,
+    AsyncUnaryCallContinuation<TRequest, TResponse> continuation)
+{
+    LogCall(context.Method);
+    AddCallerMetadata(ref context);
+
+    var call = continuation(request, context);
+
+    return new AsyncUnaryCall<TResponse>(HandleResponse(call.ResponseAsync), call.ResponseHeadersAsync, call.GetStatus, call.GetTrailers, call.Dispose);
+}
+```
```

Comment:
This is a pretty complex interceptor (adding metadata by replacing the context, handling async response), and it has a lot of methods that aren't shown. I think the most understandable example would:
- include the top-level Interceptor being inherited from.
- only show logging the call

Something like:
```cs
public class ClientLoggingInterceptor : Interceptor
{
    private readonly ILogger _logger;

    public LoggingInterceptor(ILoggerFactory loggerFactory)
    {
        _logger = loggerFactory.CreateLogger<LoggingInterceptor>();
    }

    public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(
        TRequest request,
        ClientInterceptorContext<TRequest, TResponse> context,
        AsyncUnaryCallContinuation<TRequest, TResponse> continuation)
    {
       _logger.LogInformation($"Starting call. Type: {method.Type}. Request: {typeof(TRequest)}. Response: {typeof(TResponse)}");
        return continuation(request, context);
    }
}
```
I think a more complex example can be added later. Using async is a common question and the more complex example can discuss how to via overriding `ResponseAsync`